File: 060.Permutation_Sequence/Solution.java
class Solution {
    public String getPermutation(int n, int k) {
        List<Integer> nums = new ArrayList<Integer>();
        int factorial = 1;
        for (int i = 0; i < n; i++) {
            nums.add(i + 1);
            if (i != 0) {
                factorial *= i;
            }
        }
        factorial *= n; //先求出 n 的阶乘
        StringBuilder ans = new StringBuilder();
        k = k - 1; // k 变为 k - 1
        for (int i = n; i > 0; i--) { 
            factorial /= (nums.size()); //更新为 n - 1 的阶乘
            int groupNum = k / factorial;
            int num = nums.get(groupNum);
            nums.remove(groupNum);
            k = k % factorial;
            ans.append(num);  

        }
        return ans.toString();
    }
}

File: 060.Permutation_Sequence/Solution.py
class Solution(object):
    def getPermutation(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: str
        """
        #阶乘
        def factorial(n):
            if n == 0:
                return 1
            else:
                return n * factorial(n - 1)

        def dfs(remain, s, n, k):
            if n==1:
                return s+remain[0]

            div = factorial(n-1)
            ind = k//div
            k = k%div
            return dfs(remain[:ind]+remain[ind+1:], s+remain[ind],n-1,k)
            
        return dfs([str(i) for i in range(1,n+1)],'',n,k-1)